### PR TITLE
[4.0] A zero cost huge performance win

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
- Add a small script to the arsenal of NPM tools so that scripts and stylesheets can be gzipped

### WHY?
Performance is really crucial these days, especially since Google is accounting it as a metric to rank the pages. Joomla is not meeting the today needs. Prove that in the HTTP archive the Joomla websites rank second to last (WIX): 

Link: https://discuss.httparchive.org/t/cms-performance/1468

The main reason for this is that Joomla has ALL the scripts in the head of the document and also a huge CSS file that blocks the rendering. But these cannot seem to be ok to be fixed for various reasons. So let's try to get some small wins from various tweaks.
In this PR we provide a way to gzip the script and stylesheet files ONCE (probably in the build of the installable zip files). This will free up some processor cycles in every request since the files are already gzipped so Apache or Nginx don't have to stupidly repeat the process for every request. Simple, efficient, performant. This has been my way of handling the assets for many years with 0 problems, so it's also battle tested somehow.

### How to
Run the command 
```bash
node build.js --build-gzip
```
in the root of your Joomla installation.

If you're in Apache follow the instructions here: https://httpd.apache.org/docs/2.4/mod/mod_deflate.html#page-header
Basically: 
```
<IfModule mod_headers.c>
    # Serve gzip compressed CSS files if they exist 
    # and the client accepts gzip.
    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
    RewriteCond "%{REQUEST_FILENAME}\.gz" -s
    RewriteRule "^(.*)\.css" "$1\.css\.gz" [QSA]

    # Serve gzip compressed JS files if they exist 
    # and the client accepts gzip.
    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
    RewriteCond "%{REQUEST_FILENAME}\.gz" -s
    RewriteRule "^(.*)\.js" "$1\.js\.gz" [QSA]


    # Serve correct content types, and prevent mod_deflate double gzip.
    RewriteRule "\.css\.gz$" "-" [T=text/css,E=no-gzip:1]
    RewriteRule "\.js\.gz$" "-" [T=text/javascript,E=no-gzip:1]


    <FilesMatch "(\.js\.gz|\.css\.gz)$">
      # Serve correct encoding type.
      Header append Content-Encoding gzip

      # Force proxies to cache gzipped & 
      # non-gzipped css/js files separately.
      Header append Vary Accept-Encoding
    </FilesMatch>
</IfModule>
```

### Documentation Changes Required
To production team: If you accept this you need to run the command above before each realise (can be automated, I guess) and also you need to provide a solid guide based on the info above so Admins can tweak the apache/nginx configurations. Also one thing that might needs to be considered is that the size of the installable/updates will increase by a good amount of Mbs.
Lastly if this will ever be consider I also propose that you force the extensions in the JED to also provide gzipped files for all their scripts/stylesheets.

And a final word: even if you reject to actually formally follow the proposed path, you can at least provide the tool so people that want to provide decent experiences at least have the means.

### CODE REVIEW

**Comment by richard67** [comment]
@dgrammatiko I have done this method too before my hoster finally enabled mod_deflate for shared hosting, but I did not use extension `gz`, I used `jgz` for reasons stated here:

- [https://stackoverflow.com/questions/1235116/safari-and-gzip](https://stackoverflow.com/questions/1235116/safari-and-gzip)
- [http://blog.bigsmoke.us/2012/01/16/safari-ignores-content-type-for-gz-suffix](http://blog.bigsmoke.us/2012/01/16/safari-ignores-content-type-for-gz-suffix)

I suggest that you check the information linked above, and if still relevant consider to use `jgz` instead of `gz`.

**Comment by dgrammatiko** [comment]
I don't think any of these apply here:
- safari v4 is outdated
- I'm not proposing to call/use the file.js.gz directly, this is done automagically by either Apache or Nginx.

Some explanation how all this works:
- User makes a request to our site
- Joomla creates the html and sends it back
- Browser parses the page, finds all the assets and starts making requests
- Apache receives the request
- Checks if the header that is send by the browser got the gzip deflate : eg `Accept-Encoding: br, gzip, deflate`
- If gzip is supported **apache** through the snippet provided above will switch off the automatic gzipping of the file and serve the already stored one.

The process power released by not having to gzip files in every request is enough to raise the maximum number of concurrent client. Not to mention that this is **environmentally friendly programmingⓇ**

BTW the provided default snippet by apache will **always** switch the extensions, I guess this needs an if statement to check if the file actually exists...

**Comment by richard67** [comment]
> I'm not proposing to call/use the file.js.gz directly, this is done automagically by either Apache or Nginx.
I know.

> Some explanation how all this works:
I know.

> BTW the provided default snippet by apache will always switch the extensions, I guess this needs an if statement to check if the file actually exists...

Yes, there should be a `RewriteCond %{REQUEST_FILENAME}.gz -f` at the end of the rewrite conditions chain for each of the file types.

**Comment by richard67** [comment]
Oh, I just see there is already a `RewriteCond "%{REQUEST_FILENAME}\.gz" -s`. That should do already the job.

**Comment by dgrammatiko** [comment]
Here is my nginx conf:
```
# gzip
gzip_static  on;
gzip on;
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_types text/plain text/css text/xml application/json application/javascript application/xml+rss application/atom+xml image/svg+xml application/font-woff  application/font-woff2;
```

The only extra part needed to work is `gzip_static  on;`
Ref: http://nginx.org/en/docs/http/ngx_http_gzip_static_module.html

**Comment by ggppdk** [comment]
very nice to see this pr,

i understand how this works,
i have used it myself in some cases

but i want to ask regarding the case of having apache as web server
why this PR does not include the suggested changes inside htaccess.txt too ?
i mean htaccess.txt is used in new installations so why not include the suggested modification ?
are we expecting issues with contents of 
```
<IfModule mod_headers.c>
    # Serve gzip compressed CSS files if they exist 
    # and the client accepts gzip.
...
</IfModule>
```

**Comment by dgrammatiko** [comment]
> why this PR does not include the suggested changes inside htaccess.txt too 

This PR only provides the tool for automating the creation of the .gz files. The steps to roll this to production are not part of this PR. As I wrote above the documentation is vital here and obviously tweaking the htaccess.txt and the build script are also required steps.

**Comment by Cyrusxxx** [comment]
@dgrammatiko you are a genious!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23385">issues.joomla.org/tracker/joomla-cms/23385</a>.</sub>

**Comment by dgrammatiko** [comment]
@wilsonge can we get a decision from the production team on this one?

PS. The drone is failing, probably because the linux is 32bit and the zopfli script exists only for 64bit systems.

By the way, for anyone care to dig into the used technology here that is zopfli. Zopfli is compression scripts from Google, basically an improved gzip: https://github.com/google/zopfli

**Comment by dgrammatiko** [comment]
Redid this using WASM so no further changes required.
Attention this is now based on the new code for the build tools...

**Comment by HLeithner** [comment]
is there a reason why we don't create brotli compressed files? gzip is done by the moste webservers/loadbalancers/caches on the fly but brotli support is still very limited.

**Comment by dgrammatiko** [comment]
@HLeithner we could do that as well. The idea here is that these are repeated tasks from the web server which require processor power, so by providing them ready for delivery it's only a case of reading the file and streaming it...

FWIW I like the idea that Joomla will also provide prepared Brotli files

**Comment by dgrammatiko** [comment]
@HLeithner it seems zopfli is able to achieve the same level of compression as brotli: 
screenshot 2019-02-03 at 22 42 00

**Comment by wilsonge** [comment]
@dgrammatiko you've accidentally committed the templates compiled css (and minified css) files here

**Comment by dgrammatiko** [comment]
@wilsonge can you try to build (locally) a 4.0-dev release so you can verify that this is actually working?

File: build/build.php [hunk 1]
```diff
@@ -204,8 +204,12 @@ function clean_composer(string $dir)
 	exit(1);
 }
 
+system('npm install @gfx/zopfli iltorb --unsafe-perm', $npmReturnCode);
 system('npm install --unsafe-perm', $npmReturnCode);
 
+// Create gzipped version of the static assets
+system('npm run gzip');
```
Comment (wilsonge) [inline build/build.php]: This should go after the npm install return code check

File: build/build.php [hunk 2]
```diff
@@ -204,8 +204,12 @@ function clean_composer(string $dir)
 	exit(1);
 }
 
+system('npm install @gfx/zopfli iltorb --unsafe-perm', $npmReturnCode);
```
Comment (wilsonge) [inline build/build.php]: these should be dev dependencies. All our npm stuff needs to end up in the package-lock

Comment (dgrammatiko) [reply build/build.php]: So my idea here was that this is a script that will only be executed by the dev that will create a release thus it makes 0 sense to bloat the rest of the package.json with 2 more dependencies...
I'm trying to keep npm as fast as possible so ppl with low end devices don't have to wait too long whenever they run npm install

Anyways if you think that this is not a good idea I can reverse it

Comment (mbabker) [reply build/build.php]: It's better if running from git is as close to running from a production package as possible.  The more differences there are between a git install and a ZIP install, the more potentials there are for problems to exist in release packages that won't exist in the repo.  And there are already enough of those headaches in 3.x with everything being in the repo (ref the releases where a file was added at a spot that was previously deleted and the deleted file wasn't removed from the deleted files array as a prime example).

File: build/build.php (hunk 1 above)
Comment (dgrammatiko) [reply build/build.php]: done

File: build/build.php (hunk 2 above)
Comment (dgrammatiko) [reply build/build.php]: Ok, let's go back to the dependencies

**Comment by wilsonge** [comment]
```
Processing: /Users/george/Sites/joomla-cms/media/vendor/webcomponentsjs/js/webcomponents-sd.min.js
```

After this file it hung. I had to command + c to close the task

**Comment by dgrammatiko** [comment]
@wilsonge it didn't hang it just takes very long to make all those .gz files. Try adding a console.log to see the files processing

PS the compression from zopfli is slow but the ratio is the same as brotli:
screenshot 2019-02-28 at 22 26 33

**Comment by dgrammatiko** [comment]
And for comparison here are the minified, normal gzip from a server, and the zopfli sizes:

56.658 bytes, 21.462 bytes, 17.881 bytes

screenshot 2019-02-28 at 22 42 50

**Comment by wilsonge** [comment]
Seems good to me. Thanks!

**Pull request merged**